contrib/runners/winrm_runner/winrm_runner/winrm_base.py

```diff
@@ -0,0 +1,301 @@
+        self._username = self.runner_parameters[RUNNER_USERNAME]
+        self._password = self.runner_parameters[RUNNER_PASSWORD]
+        self._timeout = self.runner_parameters.get(RUNNER_TIMEOUT, DEFAULT_TIMEOUT)
+        self._read_timeout = self._timeout + 1  # read_timeout must be > operation_timeout
+
+        # default to https port 5986 over ntlm
+        self._port = self.runner_parameters.get(RUNNER_PORT, DEFAULT_PORT)
+        self._scheme = self.runner_parameters.get(RUNNER_SCHEME, DEFAULT_SCHEME)
+        self._transport = self.runner_parameters.get(RUNNER_TRANSPORT, DEFAULT_TRANSPORT)
+
+        # if connecting to the HTTP port then we must use "http" as the scheme
+        # in the URL
+        if self._port == WINRM_HTTP_PORT:
+            self._scheme = "http"
+
+        # construct the URL for connecting to WinRM on the host
+        self._winrm_url = "{}://{}:{}/wsman".format(self._scheme, self._host, self._port)
+
+        # default to verifying SSL certs
+        self._verify_ssl = self.runner_parameters.get(RUNNER_VERIFY_SSL, DEFAULT_VERIFY_SSL)
+        self._server_cert_validation = "validate" if self._verify_ssl else "ignore"
+
+        # additional parameters
+        self._cwd = self.runner_parameters.get(RUNNER_CWD, None)
+        self._env = self.runner_parameters.get(RUNNER_ENV, {})
+        self._env = self._env or {}
+        self._kwarg_op = self.runner_parameters.get(RUNNER_KWARG_OP, DEFAULT_KWARG_OP)
+
+    def _create_session(self):
+        # create the session
+        LOG.info("Connecting via WinRM to url: {}".format(self._winrm_url))
+        session = Session(self._winrm_url,
+                          auth=(self._username, self._password),
+                          transport=self._transport,
+                          server_cert_validation=self._server_cert_validation,
+                          operation_timeout_sec=self._timeout,
+                          read_timeout_sec=self._read_timeout)
+        return session
+
+    def _winrm_get_command_output(self, protocol, shell_id, command_id):
+        # NOTE: this is copied from pywinrm because it doesn't support
+        # timeouts
+        stdout_buffer, stderr_buffer = [], []
+        return_code = 0
+        command_done = False
+        start_time = time.time()
+        while not command_done:
+            # check if we need to timeout (StackStorm custom)
+            current_time = time.time()
+            elapsed_time = (current_time - start_time)
+            if self._timeout and (elapsed_time > self._timeout):
+                raise WinRmRunnerTimoutError(Response((b''.join(stdout_buffer),
+                                                       b''.join(stderr_buffer),
+                                                       WINRM_TIMEOUT_EXIT_CODE)))
+            # end stackstorm custom
+
+            try:
+                stdout, stderr, return_code, command_done = \
+                    protocol._raw_get_command_output(shell_id, command_id)
```

Comment:
How does this function behave? Does it block until it receives output / data? If so, for how long? Is that user configurable?

Is it possible that this function returns immediately in some situations? In this case we probably need to add some `eventlet.sleep(0.2)` or similar to prevent blocking / busy waiting.